### PR TITLE
perf(model): Replace Sklearn with optimized PyTorch KMeans in CFA

### PR DESCRIPTION
**Pull request by andersendsa**
## 📝 Description
Addresses TODO in `src/anomalib/models/image/cfa/torch_model.py` to use a PyTorch KMeans class instead of Sklearn's version which requires moving data to the CPU. - Uses `anomalib.models.components.cluster.kmeans.KMeans` and optimizes it using PyTorch's vectorized `bincount` and `index_add_` operations instead of python loops with boolean indexing, providing significant performance speedups.

## ✨ Changes

Select what type of change your PR is:

- [ ] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🔄 Refactor (non-breaking change which refactors the code base)
- [x] ⚡ Performance improvements
- [ ] 🎨 Style changes (code style/formatting)
- [ ] 🧪 Tests (adding/modifying tests)
- [ ] 📚 Documentation update
- [ ] 📦 Build system changes
- [ ] 🚧 CI/CD configuration
- [ ] 🔧 Chore (general maintenance)
- [ ] 🔒 Security update
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)

## ✅ Checklist

Before you submit your pull request, please make sure you have completed the following steps:

- [ ] 📚 I have made the necessary updates to the documentation (if applicable).
- [ ] 🧪 I have written tests that support my changes and prove that my fix is effective or my feature works (if applicable).
- [ ] 🏷️ My PR title follows conventional commit format.

For more information about code review checklists, see the [Code Review Checklist](https://github.com/open-edge-platform/anomalib/blob/main/docs/source/markdown/guides/developer/contributing.md).

### CODE REVIEW

File: src/anomalib/models/image/cfa/torch_model.py [hunk 1]
```diff
@@ -240,18 +240,13 @@ def initialize_centroid(self, data_loader: DataLoader) -> None:
         scale = self.get_scale(batch.shape[-2:])
 
         if self.gamma_c > 1:
-            # TODO(samet-akcay): Create PyTorch KMeans class.
             # CVS-122673
```
Comment (ashwinvaidya17) [inline src/anomalib/models/image/cfa/torch_model.py]: maybe we can remove the cvs as well

**Review by ashwinvaidya17 — CHANGES_REQUESTED** [review]
Thanks for updating this. I have left a very minor comment. Can you also have a look at the copilot comments as well. They are quite minor as well

**Comment by ashwinvaidya17** [comment]
Also, can you update the PR title

File: src/anomalib/models/components/cluster/kmeans.py [hunk 2]
```diff
@@ -93,10 +93,12 @@ def fit(self, inputs: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
...
-            for j in range(self.n_clusters):
-                mask = self.labels_ == j
-                if mask.any():
-                    self.cluster_centers_[j] = inputs[mask].mean(dim=0)
+            counts = torch.bincount(self.labels_, minlength=self.n_clusters).to(inputs.dtype)
+            new_centers = torch.zeros_like(self.cluster_centers_)
+            new_centers.index_add_(0, self.labels_, inputs)
+
+            valid_mask = counts > 0
+            self.cluster_centers_[valid_mask] = new_centers[valid_mask] / counts[valid_mask].unsqueeze(1)
```
[the remark this answers is not in the document]
Comment (andersendsa) [reply src/anomalib/models/components/cluster/kmeans.py:101]: implemented

File: src/anomalib/models/image/cfa/torch_model.py [hunk 3]
```diff
@@ -49,6 +48,7 @@
 
 from anomalib.data import InferenceBatch
 from anomalib.models.components import DynamicBufferMixin
+from anomalib.models.components.cluster.kmeans import KMeans
 from anomalib.models.components.feature_extractors import dryrun_find_featuremap_dims
```
[the remark this answers is not in the document]
Comment (andersendsa) [reply src/anomalib/models/image/cfa/torch_model.py:52]: implemented

File: src/anomalib/models/image/cfa/torch_model.py (hunk 1 above)
Comment (andersendsa) [reply src/anomalib/models/image/cfa/torch_model.py]: have done the change

File: src/anomalib/models/components/cluster/kmeans.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (andersendsa) [reply src/anomalib/models/components/cluster/kmeans.py:101]: implemented

**Comment by andersendsa** [comment]
Hi @ashwinvaidya17  I have added unit tests could you pls let me know if this pr needs any changes or if it is good to merge

**Comment by ashwinvaidya17** [comment]
@andersendsa I know continuous copilot comments can be annoying but can you check if they make sense and update the PR if necessary?

**Comment by andersendsa** [comment]
> @andersendsa I know continuous copilot comments can be annoying but can you check if they make sense and update the PR if necessary?

Hi @ashwinvaidya17  sorry for the delay I have implemented the changes required and resolved them pls let me know if this pr needs anymore changes or if it is good to merge

**Review by ashwinvaidya17 — COMMENTED** [review]
Thanks for the efforts. Just one minor comment. I think we should add early exit once the labels stop changing. Otherwise we might keep it running till max_iter

**Comment by andersendsa** [comment]
> Thanks for the efforts. Just one minor comment. I think we should add early exit once the labels stop changing. Otherwise we might keep it running till max_iter

Hi @ashwinvaidya17  I have done the changes pls let me know if pr needs anymore changes or if it is good to merge

**Comment by ashwinvaidya17** [comment]
@andersendsa can you fix the prek issues? I think we are good to merge after that.

**Comment by andersendsa** [comment]
> @andersendsa can you fix the prek issues? I think we are good to merge after that.

Hi @ashwinvaidya17  I have fixed the prek issues and This pr is good to merge
